tests: mem_map: fix execution test for x86_64 with coverage

dcpleung opened the pull request:
When coverage is enabled on x86_64, GCC uses relative addressing
to increment the gcov counters. The generated code of the test
function assumes execution is in the same location where
the linker places the test function. This does not work with
the execution test as it copies the function into another part
of memory and tries to execute from there. When the copied
function starts to run, the instruction pointer is at the newly
copied function. So any relative addressing with regard to
the instruction pointer now is invalid. Instead of
<generated code RIP + offset> for gcov counter as it should be,
now the copied code is trying to access the counter at
<copied code RIP + offset>, which points to incorrect
memory location (and possibly invalid/non-mapped memory).
To fix this, we need to tell GCC not to use relative addressing.
This can be accomplished by telling GCC to use the large memory
model. This is only used for this test as this option increases
code size quite a bit, and should not be used in general.

Fixes #30434

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: mem_map: fix execution test for x86_64 with coverage

When coverage is enabled on x86_64, GCC uses relative addressing
to increment the gcov counters. The generated code of the test
function assumes execution is in the same location where
the linker places the test function. This does not work with
the execution test as it copies the function into another part
of memory and tries to execute from there. When the copied
function starts to run, the instruction pointer is at the newly
copied function. So any relative addressing with regard to
the instruction pointer now is invalid. Instead of
<generated code RIP + offset> for gcov counter as it should be,
now the copied code is trying to access the counter at
<copied code RIP + offset>, which points to incorrect
memory location (and possibly invalid/non-mapped memory).
To fix this, we need to tell GCC not to use relative addressing.
This can be accomplished by telling GCC to use the large memory
model. This is only used for this test as this option increases
code size quite a bit, and should not be used in general.

Fixes #30434

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

andrewboie commented:
I would feel better if we could also guarantee that the function's code does not cross a page boundary -- this is unlikely but possible. The test copies the page containing the function's starting address, but if the entire function spans a page there will be a crash.

dcpleung commented:
There currently has no mechanism to do a app only linker script. So it's hard to know the size of the function without with start/end linker magic.

andrewboie commented:
> There currently has no mechanism to do a app only linker script.

You positive? samples/application_development/code_relocation uses a custom `zephyr_linker_sources(SECTIONS custom-sections.ld)`

The function will never be larger than a page, so if we can get its base address aligned to a page it should be sufficient.

dcpleung commented:
Ah... I was only searching for `*.ld` in under `tests/` and not `'samples/` so I didn't see this. Let me take a look at it.

dcpleung committed:
tests: mem_map: put transplanted_function into its own section

This puts the transplanted_function into its own section so that
z_phys_map() can correctly map the whole range of memory used
by the function, in case someone decides to expand the function
to be bigger than a MMU page.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Added another commit to move the function into its own section.

andrewboie commented on tests/kernel/mem_protect/mem_map/testcase.yaml:
```diff
@@ -2,3 +2,17 @@ tests:
     filter: CONFIG_MMU
+    extra_sections: _TRANSPLANTED_FUNC
+    platform_exclude: qemu_x86_64
+  kernel.memory_protection.mem_map.qemu_x86_64:
```
not sure why we need 2 scenarios for x64?

dcpleung replied:
Because changing the memory model also changes how code is generated. This is an attempt to make sure that tests other than the execution test still test code generated when building without the compiler switch.
